### PR TITLE
Add Rails 7 Support

### PR DESCRIPTION
**Pull request by tagliala**
# ⚠️  Do not rely on this branch for deployment ⚠️

Hi, this is a work in progress and I'm opening this PR as a reference to #7196 

I'm trying to update activeadmin to use Rails 7, but there are some blocking dependencies:

### ✅  No known issues
- [x] Incompatible with Server Timing middleware: https://github.com/activeadmin/activeadmin/issues/7196#issuecomment-1002878544 #7262
- [x] Autoloading issues when `before_action` option is enabled:  https://github.com/activeadmin/activeadmin/issues/7196#issuecomment-1004789728 https://github.com/activeadmin/activeadmin/pull/7235#issuecomment-1030803856 #7340
### ✅ Dependencies checklist

- [x] [inherited_resources](https://github.com/activeadmin/inherited_resources) activeadmin/inherited_resources#707
- [x] [ransack](https://github.com/activerecord-hackery/ransack) activerecord-hackery/ransack#1266 
- [x] [cucumber-rails](https://github.com/cucumber/cucumber-rails) cucumber/cucumber-rails#523 cucumber/cucumber-rails/pull/522 - #7366 #7369 
- [x] [rails-i18n](https://github.com/svenfuchs/rails-i18n) svenfuchs/rails-i18n#957 svenfuchs/rails-i18n#958
- [x] [db-query-matchers](https://github.com/civiccc/db-query-matchers) civiccc/db-query-matchers#42 - downgrades to an old version without issues #7232
- [x] [arbre](https://github.com/activeadmin/arbre) activeadmin/arbre#314 - downgrades to an old version **with** issues - **ISSUES FIXED, WAITING FOR RELEASE**

### ✅ Removed dependencies
- [x] [activerecord-jdbc-adapter](https://github.com/jruby/activerecord-jdbc-adapter) jruby/activerecord-jdbc-adapter#1103

At the moment, there are no stable JRuby versions supporting MRI 2.7, the minimum required version to run Rails 7.0. There was too much effort in trying to build against jruby-head, so JRuby dependency has been temporarily removed

#### 1. Changes to Gemfile

Add the following to your Gemfile:

```rb
# NOTE: DO NOT RELY ON THIS BRANCH FOR DEPLOYMENT!
gem 'activeadmin', github: 'tagliala/activeadmin', branch: 'feature/railties-7' # FIXME: revert to stable
```

### Tests

- [x] I cannot replicate the zeitwerk failure I see on CI on my local development machine (specs are green)
- [x] I have cucumber related failures because of `uninitialized constant #<Class:Cucumber::Rails::Database>::NullStrategy (NameError)`. It is worth to mention that cucumber-rails does seem to work with Rails 7, or at least it does not have specs covering this issue. Ref: https://github.com/cucumber/cucumber-rails/issues/523#issuecomment-998118841

### CODE REVIEW

**Comment by tagliala** [comment]
Updates:
- a new version of rails-i18n supporting rails 7 has been released
- cucumber features pass on Rails 7 (cucumber/cucumber-rails#522 - 7253b1ffd47cbfbc3b919f7385934b3c62462b40)
- I'm still not able to replicate the spec failure on my development machine or in a linux docker container

```
| An error occurred while loading ./spec/requests/default_namespace_spec.rb.
| Failure/Error: require "#{ActiveAdmin::TestApplication.new.full_app_dir}/config/environment.rb"
| 
| Zeitwerk::NameError:
|   expected file ~/dev/activeadmin/tmp/test_apps/rails_70/app/admin/admin_users.rb to define constant AdminUsers, but didn't
```

I can replicate this issue locally when I try to run CI workflows using [act](https://github.com/nektos/act), but if I login on the docker container and I try there, specs pass. Any hint would help

**Comment by tagliala** [comment]
### 🟢  [Specs are green!](https://github.com/tagliala/activeadmin/runs/4628320041)

I've got the reason of the failure (and - of course - it was pretty simple)
 
Rails 7's default `config/environments/test.rb` has the following change:

```
config.eager_load = ENV["CI"].present?
```

So Rails 7 is going to perform eager loading and it fails, because `remove_active_admin_load_paths_from_rails_autoload_and_eager_load` does not prevent eager loading of classes in that folder

I'm not sure of the fix (disable eager loading), please let me know what do you think

**Comment by tagliala** [comment]
### UPDATE

Ransack 2.5.0 released, four to go:

- arbre (pending release)
- cucumber-rails
- db-query-matchers (pending merge and release)
- inherited_resources (pending release)

Cucumber Rails may need help, there are two conditional test case to write --> cucumber/cucumber-rails#523

**Comment by tagliala** [comment]
### UPDATE

There is an issue with Rails 7, code reloading, and the new `server_timing` option

It is described here: https://github.com/activeadmin/activeadmin/issues/7196#issuecomment-1002878544

It can be replicated by changing something in an activeadmin file (`/app/admin/dashboard.rb`) and reloading the page

Changing `server_timing` to `false` is a workaround, but this issue should be fixed. Any help would be appreciated

**Comment by alejandroperea** [comment]
@tagliala this PR has been merged https://github.com/activeadmin/activeadmin/pull/7232. It replaces the `db-query-matchers` gem with an Rspec matcher that is doing the same check than the gem.

Thanks a lot for your effort!

**Comment by tagliala** [comment]
Hi @alejandroperea, thanks!

I was hoping for #7275 to be merged before rebasing, so I can clean up other things and remove unrelated changes and gem updates

**Comment by javierjulio** [comment]
@tagliala you're good to go as #7275 is merged. I'll approve and run CI here once you push up that update. Thanks again!

**Comment by tagliala** [comment]
>  I'll approve and run CI here once you push up that update.

There is still another issue not covered by spec, so unfortunately this can't be merged 
 https://github.com/activeadmin/activeadmin/issues/7196#issuecomment-1018619384

**Comment by javierjulio** [comment]
@tagliala no worries, I didn't mean to merge but just to get CI to run on your next push since there are merge conflicts and you mentioned you wanted to do other clean up as well. CI won't run automatically as I will have to click "Approve and run" first.

**Comment by tagliala** [comment]
There is a cache issue in github actions but I do not know how to fix it 🤷🏼 

https://github.com/activeadmin/activeadmin/runs/4914540076?check_suite_focus=true#step:5:820

**Comment by mgrunberg** [comment]
> There is a cache issue in github actions but I do not know how to fix it 🤷🏼
> 
> https://github.com/activeadmin/activeadmin/runs/4914540076?check_suite_focus=true#step:5:820

@tagliala Bundler complains about missing Gemfile.lock changes. You changed `activeadmin.gemspec` extending `railties` range. All `gemfile/*/Gemfile` refer to gemspec. So, some `gemfile/*.lock` files need updates. I made [this commit](https://github.com/mgrunberg/activeadmin/commit/9e588019210b42731d2c1ba3d0d203780e696c91) showing what I mean

**Comment by alejandroperea** [comment]
Fixing the problem with the latest builds here: https://github.com/activeadmin/activeadmin/pull/7302

File: lib/active_admin/base_controller/authorization.rb
Comment (tagliala) [inline lib/active_admin/base_controller/authorization.rb:89]: you may want to start reviewing this. There is also code duplication

I don't know if there are side effects and if it a good idea to have a single constantizer helper defined in the ActiveAdmin module so that both AuthorizationController and Resource can use it

Comment (technicalpickles) [reply lib/active_admin/base_controller/authorization.rb:89]: `String#constantize` has been available since Rails 3, so I think it's safe to use, and that removes the need for a helper.

Comment (tagliala) [reply lib/active_admin/base_controller/authorization.rb:89]: Thanks.

I've actually checked git blame and that method has been introduced in a commit 10 years ago 572dbf1ffafcdc30d41aa2146b4a49ffb8b1c32d

Will remove the helper and use constantize

Comment (tagliala) [reply lib/active_admin/base_controller/authorization.rb]: @technicalpickles I've tried to remove `ActiveSupport::Dependencies.constantize(resource_name)`, but specs are failing now, even on 7.0

There is a difference in behavior

`rspec ./spec/unit/filters/active_filter_spec.rb:220 `

```
(byebug) resource_name
"::SuperPost"
(byebug) ActiveSupport::Dependencies.constantize(resource_name)
#<Class:0x00007f9cd686fdb0>(id: integer, title: string, body: text, published_date: date, author_id: integer, position: integer, custom_category_id: integer, starred: boolean, foo_id: integer, created_at: datetime, updated_at: datetime)
(byebug) resource_name.constantize
*** NameError Exception: uninitialized constant SuperPost

nil
```

Ref: https://github.com/activeadmin/activeadmin/runs/5028708103?check_suite_focus=true

Comment (technicalpickles) [reply lib/active_admin/base_controller/authorization.rb]: Spent a little time looking at how `constantize` works, and it seems that ActiveSupport::Dependencies.constantize is pretty closely tied to classic model. The zeitwerk mode only does the `constantize` like this, so ActiveSupport::Dependencies.constantize must be doing some extra work around loading.

I think that means adding a private helper to use `ActiveSupport::Dependencies.constantize` if available, or `adapter.constantize` otherwise.

Comment (technicalpickles) [reply lib/active_admin/base_controller/authorization.rb]: Oops, I missed this is happening on Rails 7 too 🤔

Comment (tagliala) [reply lib/active_admin/base_controller/authorization.rb:89]: > Oops, I missed this is happening on Rails 7 too 🤔

Hi @technicalpickles , sorry, I think that it may be the test itself

https://github.com/activeadmin/activeadmin/blob/1b6dd5c2c5722f896ca36c3fa447b4db63038446/spec/unit/filters/active_filter_spec.rb#L221

Comment (tagliala) [reply lib/active_admin/base_controller/authorization.rb]: @technicalpickles apologies, my fault. I didn't even check the spec definition. Now it is fixed

Aside of using jsbundling, tracked in #7264,  the only known issue is the `before_action` setting

**Comment by tagliala** [comment]
The last known issue at the moment happens here:

https://github.com/activeadmin/activeadmin/blob/72336a175213104e4ef4cafd22d0bf8aa59174f3/lib/active_admin/application.rb#L160

At the moment this code is executed, BaseController is not available. If you explicitly require `inherited_resources/base`, then ApplicationController will be not available when BaseController needs it

**Comment by baarkerlounger** [comment]
@tagliala looks like Arbre 1.5.0 has been released now so can bump that here I think

**Comment by baarkerlounger** [comment]
@tagliala just hit the BaseController/Inherited Resource/Base autoloading issue. Do you have a workaround/fix already I could try?

**Comment by tagliala** [comment]
> Do you have a workaround/fix already I could try?

Yes! https://github.com/activeadmin/activeadmin/issues/7196#issuecomment-1018619384

I'm not going to propose a fix by wrapping the whole configuration in an `after_initialize` block, but it is a workaround

**Comment by tagliala** [comment]
I think that there is another development dependency that needs to be updated: [activerecord-jdbc-adapter](https://github.com/jruby/activerecord-jdbc-adapter).

I can't find on that repository an issue related to Rails 7

File: Gemfile.lock
Comment (tagliala) [inline Gemfile.lock:87]: Downgrade not expected. There are no Rails 7.0 compatible releases

Comment (deivid-rodriguez) [reply Gemfile.lock:87]: Indeed! I guess this is because this old version didn't specify a requirement on Rails, so bundler "correctly" resolves to that.

I guess the best thing we can do is fork the repository, relax the Rails requirement to allow Rails 7, point to that fork for now and let it fail in whatever way it fails. Ideally it doesn't fail at all and we can create a PR upstream to relax the requirement. Actually that PR can be probably created right away, so that people can start messing with their master branch and trying to add support.

Comment (tagliala) [reply Gemfile.lock:87]: I've noticed that `activerecord-jdbc-adapter`'s approach would be moving the current `master` branch to `61-stable` and start committing Rails 7 related PRs to `master`. I'm not confident in testing against JRuby or changing the default CI build to "7-0-stable", should I open an issue instead?

Comment (deivid-rodriguez) [reply Gemfile.lock:87]: Last time I went ahead and created a PR directly: https://github.com/jruby/activerecord-jdbc-adapter/pull/1076, I think doing that should be fine, but if you want to gather some feedback from the repository maintainers first, feel free to open an issue 👍.

Comment (tagliala) [reply Gemfile.lock:87]: Maybe this is locked upstream because of jruby/jruby#6464. As far as I can tell, there are no JRuby stable versions compatible with Rails 7

Comment (deivid-rodriguez) [reply Gemfile.lock:87]: Indeed! I think it still makes sense to propose the update to `activerecord-jdbc-adapter`. It allows us to at least have a proper "Rails 7 compatible" resolution on JRuby, and it allows testing Rails 7 support on jruby-head.

**Comment by tagliala** [comment]
> Is the branch getting rebased or commits deleted? I've not seen this before. I know the comment says not to use this for deployment, but we're using it to run tests to make sure everything works as the branch is updated.

Yes, I'm rebasing and force pushing on that branch, so bundler is not able to resolve the commit hash 

I was thinking to maintain two separate branches, but that will require more time on my side (merge conflicts) and also will not prevent to use this branch for deployment, so at the moment I suggest to make a fork and point to a branch you have control on

**Comment by deivid-rodriguez** [comment]
@tagliala We have now merged #7340, so you can rebase now. Also, `cucumber-rails 2.5.0` is out, so you can start using that. Finally, if you point to https://github.com/jruby/activerecord-jdbc-adapter/pull/1103 so that JRuby dependencies are sorted out, we might start to have something mostly mergeable!

Thanks so much for your work!

**Comment by mgrunberg** [comment]
@tagliala I made [this PR](https://github.com/tagliala/activeadmin/pull/5) add support to eager_load on CI. I'm not doing a separate PR because the problem only affects Rails 7.x

**Comment by tagliala** [comment]
@deivid-rodriguez I can see a commit in preparation of cucumber rails 2.5.0 but it looks to me that it has not been released yet

@mgrunberg thanks, will merge that PR into mine

**Comment by deivid-rodriguez** [comment]
Right, I saw that commit and assumed the release had already happened, will have to wait some more then.

**Comment by deivid-rodriguez** [comment]
@tagliala Can you point to your fork for the activerecord-jdbc-adapter gem, so that it resolves correctly on JRuby head, and remove the FIXME string from the cucumber-rails Gemfile entry (We don't need a FIXME there, Dependabot should automatically go back to the stable release once it happens, and that should fix the lint job). I think this is ready for a review, and we don't think to wait for anything else.

File: features/support/rails.rb
Comment (tagliala) [inline features/support/rails.rb:20]: @deivid-rodriguez since `cucumber-rails` 2.5 comes with a new strategy that should be required, I'm asking if I should make another PR to point `cucumber-rails` to master in all the gemfiles, so I can remove the conditional check

Comment (deivid-rodriguez) [reply features/support/rails.rb:20]: Yes, sure!

Comment (tagliala) [reply features/support/rails.rb:20]: done in #7366

**Comment by tagliala** [comment]
Is there someone with JRuby skills to continue this PR?

I cannot install `strscan` on `jruby-head`. May be related to ruby/strscan#25

I've also problems with `racc`, `io-wait`, `net-pop`, `net-imap`

**Comment by deivid-rodriguez** [comment]
@tagliala I did some juggling to workaround related issues at https://github.com/mikel/mail/pull/1472/commits/30d1819a1d15f63d6d7f972ac5277264a730ccab. But honestly, I think it's too much work for now. Let's instead remove the java platform and the `activerecord-jdbc-adapter` (and any other jruby specific gems) altogether from the Rails 7 Gemfile. We can add it back in the future once supported.

**Review by deivid-rodriguez — APPROVED** [review]
Looks good to me! 🎉

**Comment by mgrunberg** [comment]
@tagliala cucumber-rails 2.5.0 has been released. I created https://github.com/activeadmin/activeadmin/pull/7369 to upgrade master and https://github.com/tagliala/activeadmin/pull/7 to upgrade this PR in case 7369 reaches master first.

**Comment by tagliala** [comment]
@mgrunberg thanks

I can rebase if #7369 is going to be merged before this one

**Comment by deivid-rodriguez** [comment]
I don't have a preference either way, I'll let @javierjulio choose when he gets to this :)

**Comment by sebnjk** [comment]
@tagliala Thank you for your efforts getting this updated!

**Comment by tagliala** [comment]
Got a segmentation fault on 2.7 vs 6.1. Any chance to restart the build? :\

**Comment by javierjulio** [comment]
Yes I will, thanks for checking it as I was about to ask whether we needed a rebuild or not.

**Review by javierjulio — APPROVED** [review]
Thank you @tagliala! 🙌🏻❤️

**Comment by deivid-rodriguez** [comment]
Thanks so much @tagliala and @mgrunberg!!

**Comment by grantspeelman** [comment]
Fantastic Work 🥳

**Comment by dwaynemac** [comment]
This is awesome. Thank you very much

**Comment by tierra** [comment]
Were you using `classic` mode before? (as that was removed in Rails 7)

Have you checked this?

```
bin/rails zeitwerk:check
```

You might want to review the [Classic to Zeitwerk HOWTO](https://guides.rubyonrails.org/classic_to_zeitwerk_howto.html)

Anecdotally, I've finished upgrading my app (using ActiveAdmin) to Rails 7 recently, using this update, and it went smoothly. No issues, and no additional AA changes required, it just worked.

**Comment by yosiat** [comment]
Thanks for adding support for rails 7!

Is it possible to release a version for this PR ?

**Comment by tagliala** [comment]
Hi @yosiat , Rails 7 support has been released in v2.11.0

Changelog: https://github.com/activeadmin/activeadmin/releases/tag/v2.11.0

**Pull request merged**
